Dependency to assetshare.core leads to detection of several vulnerabilities

If you depend in your Maven project on "assetshare.core" and check the project with https://jeremylong.github.io/DependencyCheck/dependency-check-maven/check-mojo.html the following issues are found:
```
[ERROR] Failed to execute goal org.owasp:dependency-check-maven:6.5.3:check (default) on project xxx: 
[ERROR] 
[ERROR] One or more dependencies were identified with vulnerabilities: 
[ERROR] 
[ERROR] acs-aem-commons-bundle-4.3.4.jar: CVE-2022-28820
[ERROR] assetshare.core-2.2.2.jar: CVE-2021-21043
[ERROR] commons-beanutils-1.7.0.jar: CVE-2014-0114, CVE-2019-10086
[ERROR] dom4j-1.1.jar: CVE-2018-1000632, CVE-2020-10683
[ERROR] maven-compat-3.0.jar: CVE-2021-26291
[ERROR] maven-core-3.0.jar: CVE-2021-26291
[ERROR] maven-settings-3.0.jar: CVE-2021-26291
[ERROR] struts-core-1.3.8.jar: CVE-2016-1181, CVE-2016-1182, CVE-2014-0114, CVE-2015-0899
[ERROR] struts-tiles-1.3.8.jar: CVE-2016-1181, CVE-2016-1182, CVE-2014-0114, CVE-2015-0899
[ERROR] velocity-1.5.jar: CVE-2020-13936
[ERROR] velocity-tools-2.0.jar: CVE-2020-13959
[ERROR] xercesImpl-2.9.1.jar: CVE-2012-0881, CVE-2009-2625, CVE-2022-23437, CVE-2013-4002
[ERROR] 
[ERROR] See the dependency-check report for more details.
[ERROR] -> [Help 1]
```

It probably makes sense to not transitively expose the embedded dependencies in assetshare.core but rather declare them with scope "provided". Also depending on a Maven plugin (with lots of transitive deps) like in https://github.com/adobe/asset-share-commons/blob/8eb97f843e78c9dc8372a5a86a2a0b6dc99bb141/core/pom.xml#L145 is highly unusual.

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -140,28 +140,24 @@ Bundle-DocURL:
         <dependency>
             <groupId>com.google.code.findbugs</groupId>
             <artifactId>jsr305</artifactId>
+            <scope>provided</scope>
         </dependency>
         <dependency>
             <groupId>com.adobe.cq</groupId>
             <artifactId>core.wcm.components.core</artifactId>
         </dependency>
-        <!-- Embedded dependency -->
-        <dependency>
-            <groupId>org.apache.maven.plugins</groupId>
-            <artifactId>maven-dependency-plugin</artifactId>
-            <version>3.1.1</version>
-        </dependency>
         <dependency>
             <groupId>com.adobe.acs</groupId>
             <artifactId>acs-aem-commons-bundle</artifactId>
+            <scope>provided</scope><!-- should not transitively be exposed -->
         </dependency>
         <!-- JDK 11 -->
         <dependency>
             <artifactId>org.apache.sling.javax.activation</artifactId>
-            <version>0.1.0</version>
             <groupId>org.apache.sling</groupId>
+            <version>0.1.0</version>
         </dependency>
-        <!-- Unit test dependencies -->
+        <!-- Test dependencies -->
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
@@ -202,6 +198,7 @@ Bundle-DocURL:
             <groupId>org.apache.felix</groupId>
             <artifactId>org.apache.felix.framework</artifactId>
             <version>6.0.3</version>
+            <scope>test</scope>
         </dependency>
     </dependencies>
 </project>

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -127,20 +127,10 @@ Bundle-DocURL:
     </build>
 
     <dependencies>
-         <!-- uk.org.lidalia dependency MUST be placed above aem-sdk -->
-         <dependency>
-            <groupId>uk.org.lidalia</groupId>
-            <artifactId>slf4j-test</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>com.adobe.aem</groupId>
-            <artifactId>aem-sdk-api</artifactId>
-        </dependency>
         <dependency>
             <groupId>com.google.code.findbugs</groupId>
             <artifactId>jsr305</artifactId>
-            <scope>provided</scope>
+            <scope>provided</scope><!-- should not transitively be exposed -->
         </dependency>
         <dependency>
             <groupId>com.adobe.cq</groupId>
@@ -155,9 +145,8 @@ Bundle-DocURL:
         <dependency>
             <artifactId>org.apache.sling.javax.activation</artifactId>
             <groupId>org.apache.sling</groupId>
-            <version>0.1.0</version>
         </dependency>
-        <!-- Test dependencies -->
+        <!-- Start Test dependencies -->
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
@@ -194,11 +183,17 @@ Bundle-DocURL:
             <groupId>org.apache.sling</groupId>
             <artifactId>org.apache.sling.api</artifactId>
         </dependency>
-        <dependency>
-            <groupId>org.apache.felix</groupId>
-            <artifactId>org.apache.felix.framework</artifactId>
-            <version>6.0.3</version>
+        <!-- uk.org.lidalia dependency MUST be placed above aem-sdk -->
+         <dependency>
+            <groupId>uk.org.lidalia</groupId>
+            <artifactId>slf4j-test</artifactId>
             <scope>test</scope>
+        </dependency>
+        <!-- End Test dependencies -->
+        <!-- the aem-sdk-api should be the last dependency, as other dependencies should take precedence in case of package clashes -->
+        <dependency>
+            <groupId>com.adobe.aem</groupId>
+            <artifactId>aem-sdk-api</artifactId>
         </dependency>
     </dependencies>
 </project>

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -152,10 +152,6 @@ Bundle-DocURL:
             <artifactId>junit</artifactId>
         </dependency>
         <dependency>
-            <groupId>junit-addons</groupId>
-            <artifactId>junit-addons</artifactId>
-        </dependency>
-        <dependency>
             <groupId>org.mockito</groupId>
             <artifactId>mockito-core</artifactId>
         </dependency>
@@ -178,10 +174,6 @@ Bundle-DocURL:
         <dependency>
             <groupId>org.osgi</groupId>
             <artifactId>org.osgi.util.converter</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>org.apache.sling</groupId>
-            <artifactId>org.apache.sling.api</artifactId>
         </dependency>
         <!-- uk.org.lidalia dependency MUST be placed above aem-sdk -->
          <dependency>

--- a/core/src/test/java/com/adobe/aem/commons/assetshare/components/actions/download/impl/AsyncAssetRenditionsDownloadServletTest.java
+++ b/core/src/test/java/com/adobe/aem/commons/assetshare/components/actions/download/impl/AsyncAssetRenditionsDownloadServletTest.java
@@ -1,5 +1,14 @@
 package com.adobe.aem.commons.assetshare.components.actions.download.impl;
 
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.mock;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
 import com.adobe.aem.commons.assetshare.components.actions.impl.ActionHelperImpl;
 import com.adobe.aem.commons.assetshare.content.renditions.download.async.impl.AsyncAssetRenditionsDownloadServlet;
 import com.adobe.aem.commons.assetshare.testing.RequireAemMock;
@@ -7,17 +16,8 @@ import com.adobe.aem.commons.assetshare.util.RequireAem;
 import com.adobe.aem.commons.assetshare.util.impl.ExpressionEvaluatorImpl;
 import com.adobe.cq.dam.download.api.DownloadApiFactory;
 import com.adobe.cq.dam.download.api.DownloadService;
+
 import io.wcm.testing.mock.aem.junit.AemContext;
-import org.junit.Before;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.runner.RunWith;
-import org.mockito.junit.MockitoJUnitRunner;
-
-import javax.servlet.Servlet;
-
-import static org.junit.Assert.assertNotNull;
-import static org.mockito.Mockito.mock;
 
 @RunWith(MockitoJUnitRunner.class)
 public class AsyncAssetRenditionsDownloadServletTest {
@@ -38,9 +38,7 @@ public class AsyncAssetRenditionsDownloadServletTest {
     public void activateInCloudReady() {
         RequireAemMock.setAem(ctx, RequireAem.Distribution.CLOUD_READY, RequireAem.ServiceType.PUBLISH);
 
-        ctx.registerInjectActivateService(new AsyncAssetRenditionsDownloadServlet());
-
-        AsyncAssetRenditionsDownloadServlet servlet = (AsyncAssetRenditionsDownloadServlet) ctx.getService(Servlet.class);
+        AsyncAssetRenditionsDownloadServlet servlet = ctx.registerInjectActivateService(new AsyncAssetRenditionsDownloadServlet());
 
         assertNotNull(servlet);
     }
@@ -51,6 +49,5 @@ public class AsyncAssetRenditionsDownloadServletTest {
 
         ctx.registerInjectActivateService(new AsyncAssetRenditionsDownloadServlet());
 
-        ctx.getService(Servlet.class);
     }
 }

--- a/core/src/test/java/com/adobe/aem/commons/assetshare/content/impl/datasources/AssetRenditionsDataSourceTest.java
+++ b/core/src/test/java/com/adobe/aem/commons/assetshare/content/impl/datasources/AssetRenditionsDataSourceTest.java
@@ -102,11 +102,9 @@ public class AssetRenditionsDataSourceTest {
         String[] expected = new String[]{"a", "b", "c", "d"};
 
         ctx.currentResource("/apps/dialog/default");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET");
-
-        final Servlet servlet = ctx.getService(Servlet.class);
 
         servlet.service(ctx.request(), ctx.response());
 
@@ -121,12 +119,10 @@ public class AssetRenditionsDataSourceTest {
         String[] expected = new String[]{"c", "d"};
 
         ctx.currentResource("/apps/dialog/default");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET",
                 "exclude.assetrenditiondispatcher.names", "one");
-
-        final Servlet servlet = ctx.getService(Servlet.class);
 
         servlet.service(ctx.request(), ctx.response());
 
@@ -141,12 +137,10 @@ public class AssetRenditionsDataSourceTest {
         String[] expected = new String[]{"a", "c", "d"};
 
         ctx.currentResource("/apps/dialog/default");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET",
                 "exclude.assetrendition.names", "b");
-
-        final Servlet servlet = ctx.getService(Servlet.class);
 
         servlet.service(ctx.request(), ctx.response());
 
@@ -161,11 +155,9 @@ public class AssetRenditionsDataSourceTest {
         String[] expected = new String[]{"c", "d"};
 
         ctx.currentResource("/apps/dialog/exclude-assetrenditiondispatchers");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET");
-
-        final Servlet servlet = ctx.getService(Servlet.class);
 
         servlet.service(ctx.request(), ctx.response());
 
@@ -180,11 +172,9 @@ public class AssetRenditionsDataSourceTest {
         String[] expected = new String[]{"a", "c", "d"};
 
         ctx.currentResource("/apps/dialog/exclude-assetrenditions");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET");
-
-        final Servlet servlet = ctx.getService(Servlet.class);
 
         servlet.service(ctx.request(), ctx.response());
 
@@ -207,7 +197,7 @@ public class AssetRenditionsDataSourceTest {
         final String[] expectedValues = new String[]{"a", "b", "c", "d"};
 
         ctx.currentResource("/apps/dialog/default");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "add.assetrenditiondispatcher.to.label", true,
                 "sling.servlet.methods", "GET");
@@ -224,9 +214,6 @@ public class AssetRenditionsDataSourceTest {
                                 "c=preferred value for c"}).
                         build());
 
-
-        final Servlet servlet = ctx.getService(Servlet.class);
-
         servlet.service(ctx.request(), ctx.response());
 
         final DataSource sds = (DataSource) ctx.request().getAttribute(DataSource.class.getName());
@@ -241,7 +228,7 @@ public class AssetRenditionsDataSourceTest {
         final String[] expectedValues = new String[]{"a", "b", "c", "d"};
 
         ctx.currentResource("/apps/dialog/default");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET");
 
@@ -257,9 +244,6 @@ public class AssetRenditionsDataSourceTest {
                                 "f=value"}).
                         build());
 
-
-        final Servlet servlet = ctx.getService(Servlet.class);
-
         servlet.service(ctx.request(), ctx.response());
 
         final DataSource sds = (DataSource) ctx.request().getAttribute(DataSource.class.getName());
@@ -273,7 +257,7 @@ public class AssetRenditionsDataSourceTest {
         final String[] expectedValues = new String[]{"a", "b", "e", "f"};
 
         ctx.currentResource("/apps/dialog/allowed-assetrendition-types");
-        ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
+        final Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDataSource(),
                 "sling.servlet.resourceTypes", "asset-share-commons/data-sources/asset-renditions",
                 "sling.servlet.methods", "GET");
 
@@ -288,9 +272,6 @@ public class AssetRenditionsDataSourceTest {
                                 "e=value",
                                 "f=value"}).
                         build());
-
-
-        final Servlet servlet = ctx.getService(Servlet.class);
 
         servlet.service(ctx.request(), ctx.response());
 

--- a/core/src/test/java/com/adobe/aem/commons/assetshare/content/renditions/download/impl/AssetRenditionsDownloadServletTest.java
+++ b/core/src/test/java/com/adobe/aem/commons/assetshare/content/renditions/download/impl/AssetRenditionsDownloadServletTest.java
@@ -140,9 +140,7 @@ public class AssetRenditionsDownloadServletTest {
             return null; // void method, return null
         }).when(requestDispatcher).include(any(SlingHttpServletRequest.class), any(SlingHttpServletResponse.class));
 
-        ctx.registerInjectActivateService(new AssetRenditionsDownloadServlet());
-
-        AssetRenditionsDownloadServlet servlet = (AssetRenditionsDownloadServlet) ctx.getService(Servlet.class);
+        AssetRenditionsDownloadServlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDownloadServlet());
 
         ctx.currentResource("/content/download-servlet");
         ctx.request().setMethod("POST");
@@ -159,9 +157,7 @@ public class AssetRenditionsDownloadServletTest {
 
     @Test
     public void doPost_EmptyRenditionName() throws ServletException, IOException {
-        ctx.registerInjectActivateService(new AssetRenditionsDownloadServlet());
-
-        AssetRenditionsDownloadServlet servlet = (AssetRenditionsDownloadServlet) ctx.getService(Servlet.class);
+        AssetRenditionsDownloadServlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDownloadServlet());
 
         ctx.currentResource("/content/download-servlet");
         ctx.request().setMethod("POST");
@@ -178,9 +174,7 @@ public class AssetRenditionsDownloadServletTest {
 
     @Test
     public void doPost_InvalidAssetPath() throws ServletException, IOException {
-        ctx.registerInjectActivateService(new AssetRenditionsDownloadServlet());
-
-        AssetRenditionsDownloadServlet servlet = (AssetRenditionsDownloadServlet) ctx.getService(Servlet.class);
+        AssetRenditionsDownloadServlet servlet = ctx.registerInjectActivateService(new AssetRenditionsDownloadServlet());
 
         ctx.currentResource("/content/download-servlet");
         ctx.request().setMethod("POST");

--- a/core/src/test/java/com/adobe/aem/commons/assetshare/content/renditions/impl/AssetRenditionServletTest.java
+++ b/core/src/test/java/com/adobe/aem/commons/assetshare/content/renditions/impl/AssetRenditionServletTest.java
@@ -119,14 +119,14 @@ public class AssetRenditionServletTest {
                                 "testing2=^cq5dam\\.web\\..*"}).
                         build());
 
-        ctx.registerInjectActivateService(new AssetRenditionServlet());
+        Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionServlet());
 
         ctx.request().setMethod("GET");
         ctx.requestPathInfo().setResourcePath("/content/dam/test.png");
         ctx.requestPathInfo().setExtension("renditions");
         ctx.requestPathInfo().setSuffix("testing2/asset.rendition");
 
-        ctx.getService(Servlet.class).service(ctx.request(), ctx.response());
+        servlet.service(ctx.request(), ctx.response());
 
         Mockito.verify(assetRenditionDispatcher, Mockito.times(1)).dispatch(ctx.request(), ctx.response());
     }
@@ -142,14 +142,14 @@ public class AssetRenditionServletTest {
                                 "testing=value doesnt matter"}).
                         build());
 
-        ctx.registerInjectActivateService(new AssetRenditionServlet());
+        Servlet servlet = ctx.registerInjectActivateService(new AssetRenditionServlet());
 
         ctx.request().setMethod("GET");
         ctx.requestPathInfo().setResourcePath("/content/dam/test.png");
         ctx.requestPathInfo().setExtension("renditions");
         ctx.requestPathInfo().setSuffix("testing/download/dont-allow-this-parameter/asset.rendition");
 
-        ctx.getService(Servlet.class).service(ctx.request(), ctx.response());
+        servlet.service(ctx.request(), ctx.response());
 
         assertEquals(400, ctx.response().getStatus());
         Mockito.verify(assetRenditionDispatcher, Mockito.times(0)).dispatch(ctx.request(), ctx.response());

--- a/pom.xml
+++ b/pom.xml
@@ -475,6 +475,12 @@
                     <artifactId>nexus-staging-maven-plugin</artifactId>
                     <version>1.6.8</version>
                 </plugin>
+                <!-- checks for vulnerabilities in dependencies -->
+                <plugin>
+                    <groupId>org.owasp</groupId>
+                    <artifactId>dependency-check-maven</artifactId>
+                    <version>7.1.0</version>
+                </plugin>
             </plugins>
         </pluginManagement>
     </build>
@@ -710,6 +716,18 @@
                             </execution>
                         </executions>
                     </plugin>
+                    <plugin>
+                        <!-- checks for vulnerabilities in dependencies -->
+                        <groupId>org.owasp</groupId>
+                        <artifactId>dependency-check-maven</artifactId>
+                        <executions>
+                            <execution>
+                                <goals>
+                                    <goal>check</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                    </plugin>
                 </plugins>
             </build>
         </profile>
@@ -759,6 +777,18 @@
                                 <phase>package</phase>
                                 <goals>
                                     <goal>checkstyle</goal>
+                                </goals>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <plugin>
+                        <!-- checks for vulnerabilities in dependencies -->
+                        <groupId>org.owasp</groupId>
+                        <artifactId>dependency-check-maven</artifactId>
+                        <executions>
+                            <execution>
+                                <goals>
+                                    <goal>check</goal>
                                 </goals>
                             </execution>
                         </executions>

--- a/pom.xml
+++ b/pom.xml
@@ -241,6 +241,7 @@
                     <configuration>
                         <parallel>methods</parallel>
                         <threadCount>5</threadCount>
+                        <trimStackTrace>false</trimStackTrace>
                     </configuration>
                 </plugin>
                 <!-- Maven Failsafe Plugin -->
@@ -879,20 +880,20 @@
             <dependency>
                 <groupId>org.apache.sling</groupId>
                 <artifactId>org.apache.sling.testing.osgi-mock.junit4</artifactId>
-                <version>3.1.2</version>
+                <version>3.1.4</version>
                 <scope>test</scope>
             </dependency>
             <dependency>
                 <groupId>org.apache.sling</groupId>
                 <artifactId>org.apache.sling.testing.sling-mock.junit4</artifactId>
-                <!-- Bumping this to 2.4.0 changes behavior in how servlet.service(..) is processed, which breaks all servlet tests -->
-                <version>2.3.18</version>
+                <!-- upgrading to newer version not supported due to https://issues.apache.org/jira/browse/SLING-11362 -->
+                <version>3.1.2</version>
                 <scope>test</scope>
             </dependency>
             <dependency>
                 <groupId>io.wcm</groupId>
                 <artifactId>io.wcm.testing.aem-mock.junit4</artifactId>
-                <version>4.0.2</version>
+                <version>5.0.0</version>
                 <scope>test</scope>
             </dependency>
             <dependency>

--- a/pom.xml
+++ b/pom.xml
@@ -481,6 +481,9 @@
                     <groupId>org.owasp</groupId>
                     <artifactId>dependency-check-maven</artifactId>
                     <version>7.1.0</version>
+                    <configuration>
+                        <failBuildOnAnyVulnerability>true</failBuildOnAnyVulnerability>
+                    </configuration>
                 </plugin>
             </plugins>
         </pluginManagement>

